### PR TITLE
[feat] Mixture of Experts

### PR DESCRIPTION
**Pull request by blefaudeux**
## What does this PR do?
Implements Mixture of Experts as a simple Feedforward option. Uses the great MoE implementation from [FairScale](https://github.com/facebookresearch/fairscale/), implemented by @msbaines back in the days

This is really for fun and completeness' sake. Example usecase: [Sparse ViT](https://ai.googleblog.com/2022/01/scaling-vision-with-sparse-mixture-of.html)

TODO:
- [X] Dedicated feedforward option
- [x] Unit tests, unit tests, unit tests
- [x] microGPT demo (runs completely fine!)

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [x] Did you make sure to update the docs?
  - [ ] N/A
- [x] Did you write any new necessary tests?
  - [ ] N/A
- [x] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: examples/microGPT.py [hunk 1]
```diff
@@ -71,10 +71,12 @@ def __init__(
                     },
                 },
                 "feedforward_config": {
-                    "name": "FusedMLP",  # Use MLP if Triton is not available
+                    "name": "MixtureOfExperts",  # Use MLP if Triton is not available
```
Comment (blefaudeux) [inline examples/microGPT.py:74]: pulling in MoE becomes as simple as that (though distributed training adds another layer of complication)

**Comment by blefaudeux** [comment]
Sneak peek, MoE vs. MLP, 4 experts 1/4th the size. Not proving anything on the dense vs. MoE front, but at least the loss looks very reasonable

Screenshot from 2022-01-17 13-48-39

**Comment by blefaudeux** [comment]
Ah, needs more unit testing, thanks codecov !

File: xformers/components/feedforward/fused_mlp.py [hunk 2]
```diff
@@ -29,8 +29,6 @@ class FusedMlpConfig(FeedforwardConfig):
         class FusedMLP(Feedforward):
             """
             A MLP using fused linear layers.
-
-            .. warning: This is not currently competitive with PyTorch in terms of training speed
```
Comment (blefaudeux) [inline xformers/components/feedforward/fused_mlp.py:33]: not true anymore :D

**Comment by blefaudeux** [comment]
> # [Codecov](https://codecov.io/gh/facebookresearch/xformers/pull/181?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch) Report
> 
> > Merging [#181](https://codecov.io/gh/facebookresearch/xformers/pull/181?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch) ([da276f4](https://github.com/facebookresearch/xformers/commit/da276f4af2600419b5265b32a6d8717e42248d6f)) into [main](https://codecov.io/gh/facebookresearch/xformers/commit/04bb6c1903e77efac37c8a4809b2c9bb5c25bc0d?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch) ([04bb6c1](https://github.com/facebookresearch/xformers/commit/04bb6c1903e77efac37c8a4809b2c9bb5c25bc0d)) will **decrease** coverage by `0.22%`.
> > The diff coverage is `94.28%`.
> 
> [Impacted file tree graph](https://codecov.io/gh/facebookresearch/xformers/pull/181?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch)
> 
> ```diff
> @@            Coverage Diff             @@
> ##             main     #181      +/-   ##
> ==========================================
> - Coverage   90.58%   90.36%   -0.23%     
> ==========================================
>   Files          56       57       +1     
>   Lines        2837     2907      +70     
> ==========================================
> + Hits         2570     2627      +57     
> - Misses        267      280      +13     
> ```
> 
> Flag 	Coverage Δ 	
> Python 	`90.36% <94.28%> (-0.23%)` 	arrow_down
> 
> Flags with carried forward coverage won't be shown. [Click here](https://docs.codecov.io/docs/carryforward-flags?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch#carryforward-flags-in-the-pull-request-comment) to find out more.
> [Impacted Files](https://codecov.io/gh/facebookresearch/xformers/pull/181?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch) 	Coverage Δ 	
> [xformers/components/feedforward/fused_mlp.py](https://codecov.io/gh/facebookresearch/xformers/pull/181/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch#diff-eGZvcm1lcnMvY29tcG9uZW50cy9mZWVkZm9yd2FyZC9mdXNlZF9tbHAucHk=) 	`91.30% <ø> (ø)` 	
> [...rmers/components/feedforward/mixture_of_experts.py](https://codecov.io/gh/facebookresearch/xformers/pull/181/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch#diff-eGZvcm1lcnMvY29tcG9uZW50cy9mZWVkZm9yd2FyZC9taXh0dXJlX29mX2V4cGVydHMucHk=) 	`94.28% <94.28%> (ø)` 	
> [xformers/sparse/_csr_ops.py](https://codecov.io/gh/facebookresearch/xformers/pull/181/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch#diff-eGZvcm1lcnMvc3BhcnNlL19jc3Jfb3BzLnB5) 	`85.33% <0.00%> (-12.00%)` 	arrow_down
> 
> [Continue to review full report at Codecov](https://codecov.io/gh/facebookresearch/xformers/pull/181?src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch).
> 
> > **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch)
> > `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> > Powered by [Codecov](https://codecov.io/gh/facebookresearch/xformers/pull/181?src=pr&el=footer&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch). Last update [04bb6c1...da276f4](https://codecov.io/gh/facebookresearch/xformers/pull/181?src=pr&el=lastupdated&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=facebookresearch).

looks like this is wrong, the coverage loss is on _csr_ops, which is not changed..

File: xformers/components/feedforward/mixture_of_experts.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+                    dist.Backend.NCCL
+                    if torch.cuda.is_available()
+                    else dist.Backend.GLOO
+                )
+                dist.init_process_group(
+                    backend=backend,
+                    rank=0,
+                    world_size=1,
+                    init_method=init_url,
+                )
```
Comment (fmassa) [inline xformers/components/feedforward/mixture_of_experts.py:118]: I wouldn't recommend initiating the distributed mode inside a module.
This goes against current practices in PyTorch, which puts the burden on the user to do it.

Sure, it might make it easier for testing to add it here, but IMO this hides important things for the user

File: xformers/components/feedforward/mixture_of_experts.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                else:
+                    number_of_local_experts = 1
+
+            # Programatically handle the gating technique
+            gate_constructor = {
+                GateConfig.RoundRobin: RoundRobinGate,
+                GateConfig.Top2: Top2Gate,
+            }[gate_config]
+
+            self.gate = gate_constructor(dim_model, number_of_experts)
```
Comment (fmassa) [inline xformers/components/feedforward/mixture_of_experts.py:136]: IMO, this unnecessarily limits the scope of options we can provide to the model (only the two options available can be used).

I would have gone with the approach from `fairscale`, which takes a generic `nn.Module` as input. But again, this makes it necessary to use recursive instantiation.

File: xformers/components/feedforward/mixture_of_experts.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+                assert expert_constructor is not None
+
+            local_experts = torch.nn.ModuleList(
+                [expert_constructor() for _ in range(number_of_local_experts)]
+            )
+
+            self.moe = MOELayer(gate=self.gate, experts=local_experts, group=group)
+
+            self.requires_cuda = True
```
Comment (fmassa) [inline xformers/components/feedforward/mixture_of_experts.py:158]: I'm missing context here, is this used somewhere?

File: xformers/components/feedforward/mixture_of_experts.py (hunk 3 above)
Comment (blefaudeux) [reply xformers/components/feedforward/mixture_of_experts.py:118]: I was wondering when writing that (hence the logging.warning()), fine by me

File: xformers/components/feedforward/mixture_of_experts.py (hunk 4 above)
Comment (blefaudeux) [reply xformers/components/feedforward/mixture_of_experts.py:136]: ah good point, we can have all I think (same as fairscale), Union[nn.Module, Enum], best of both world, I'll update

File: xformers/components/feedforward/mixture_of_experts.py (hunk 5 above)
Comment (blefaudeux) [reply xformers/components/feedforward/mixture_of_experts.py:158]: it's an "old" flag, makes it easier for CI to test something or not test it depending on the HW needs  without maintaining an escape list in different places (I think that it came from the Triton parts). We can change that, it was "a" way to solve this

File: xformers/components/feedforward/mixture_of_experts.py (hunk 3 above)
Comment (blefaudeux) [reply xformers/components/feedforward/mixture_of_experts.py]: done

File: xformers/components/feedforward/mixture_of_experts.py (hunk 5 above)
Comment (blefaudeux) [reply xformers/components/feedforward/mixture_of_experts.py:145]: done

File: xformers/components/feedforward/mixture_of_experts.py (hunk 4 above)
Comment (blefaudeux) [reply xformers/components/feedforward/mixture_of_experts.py]: done

**Review by dianaml0 — APPROVED** [review]
LGTM!

**Comment by blefaudeux** [comment]
rebased, conflict resolved
